policy/k8s: Add support for CIDRGroupRef in IngressDeny and EgressDeny

pippolo84 opened the pull request:
Current version of CNP translation lacks support for translating referenced CiliumCIDRGroup objects in IngressDeny rules, despite mentioning it in the CIDRGroupRef field docstring.

The commit adds the missing logic and extends the unit tests suite to take into account the IngressDeny rules.

pippolo84 committed:
policy/k8s: Add support for CIDRGroupRef in IngressDeny

Current version of CNP translation lacks support for translating
referenced CiliumCIDRGroup objects in IngressDeny rules, despite
mentioning it in the CIDRGroupRef field docstring.

The commit adds the missing logic and extends the unit tests suite to
take into account the IngressDeny rules.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

joestringer commented on pkg/policy/k8s/cilium_cidr_group.go:
FWIW I appreciate that this is a small incremental change to add a feature that is adjacent to existing code, but this approach is continuing to complicate the policy ingestion pipeline in ways that I think are diverging from the long term correct solution for CiliumCIDRGroup, as I expressed in my [original feedback](https://github.com/cilium/cilium/pull/24638#pullrequestreview-1373839746) back a year ago.

pippolo84 replied:
Is [this](https://github.com/cilium/cilium/pull/24638#discussion_r1159194073) limitation still in place? AFAIU that was the primary blocker for moving to an ipcache-based approach.

joestringer replied:
I've summarized the outline of the proposal and latest status here: https://github.com/cilium/cilium/issues/30992 . There was some progress as recently as last week around a bitlpm data structure.

pippolo84 replied:
So, I briefly discussed this offline with @gandro .
I agree we should aim to refactor the CIDRGroupRef with a dedicated LabelSelector for the `cidrgroup:name` label, but since that requires an extensive effort compared to this solution, I think we should still consider going ahead with this, in the meantime. As you mentioned, this is just an incremental change based on the existing logic, so I think it is still acceptable while we work on the proper implementation.
I'll leave the PR open for reviews, let me know if you prefer to reconsider this.

joestringer replied:
How do we avoid this happening again?

I think it's very easy for a project to focus on features,features,features and fail to dedicate time to technical debt. This leads to incompatibility between features and weird bugs over time. We as contributors all suffer from the impacts of this, and of course our users suffer the impacts as well. It's one thing to say that we have a disagreement on the approach for this solution -- I'd be happy to have that debate. But it seems like we all agree on the long term solution, but no-one plans to actually work on it?

I don't like to be in the position of considering whether to block a useful PR like this, but this is the last opportunity we seem to have to raise these concerns and force contributors to do something about it. I feel like we should be having these discussions openly in @cilium/sig-policy earlier and planning for how to address these issues over time so that we don't end up in this position in the first place.

joestringer replied:
I guess taking a step back from this, taking some inspiration from [Google Eng Practices](https://google.github.io/eng-practices/review/reviewer/standard.html) philosophy of code review:

> In general, reviewers should favor approving a CL once it is in a state where it definitely improves the overall code health of the system being worked on, even if the CL isn’t perfect. 

The changes in this PR seem generally useful and there's a set of test cases to validate the new codepaths. The changes aren't introducing new structural complexity or so on, it's a natural extension of the existing code. I don't want to punish good quality contributions, my goal is mainly just to normalize talking about how we maintain quality and clean up the tree as we continue to develop new changes. 

With that in mind I'll resolve the comment thread here and defer to the reviewers of the specific codeowner areas.

gandro commented on pkg/policy/k8s/cilium_cidr_group.go:
```diff
@@ -175,11 +182,15 @@ func getCIDRGroupRefs(cnp *types.SlimCNP) []string {
+				}
+			}
+		}
 		for _, egress := range spec.Egress {
```
Any reason we don't also check EgressDeny?

pippolo84 replied:
Not really, just that https://github.com/cilium/cilium/pull/30624 was opened after this one.
I was thinking about doing that in a follow up PR, but let's do it here, it is easy enough. :+1:

pippolo84 replied:
Added support for EgressDeny too. Updated tests and docs :+1:

pippolo84 committed:
policy/k8s: Add support for CIDRGroupRef in EgressDeny

Current version of CNP translation lacks support for translating
referenced CiliumCIDRGroup objects in EgressDeny rules.

The commit adds the missing logic and extends the unit tests suite to
take into account the EgressDeny rules.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
policy: Mention EgressDeny in CIDRGroupRef docs

Update CIDRGroupRef docstring to take into account the support for
referenced CIDRGroup in EgressDeny rules.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro reviewed (approved):
Looks good to me now, thanks!
